<a href="https://colab.research.google.com/github/AkshayaNemalipuri/Capstone_MedIntel/blob/main/MedIntel_AI_Model_Development.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("/content/diabetes_prediction_dataset.csv")

In [4]:
df.head()


,gender,age,hypertension,heart_disease,smoking_history,bmi,HbA1c_level,blood_glucose_level,diabetes
0,Female,80.0,0,1,never,25.19,6.6,140,0
1,Female,54.0,0,0,No Info,27.32,6.6,80,0
2,Male,28.0,0,0,never,27.32,5.7,158,0
3,Female,36.0,0,0,current,23.45,5.0,155,0
4,Male,76.0,1,1,current,20.14,4.8,155,0


In [5]:
df.shape

(100000, 9)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   gender               100000 non-null  object 
 1   age                  100000 non-null  float64
 2   hypertension         100000 non-null  int64  
 3   heart_disease        100000 non-null  int64  
 4   smoking_history      100000 non-null  object 
 5   bmi                  100000 non-null  float64
 6   HbA1c_level          100000 non-null  float64
 7   blood_glucose_level  100000 non-null  int64  
 8   diabetes             100000 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 6.9+ MB


In [7]:
df.isnull().sum()

,0
gender,0
age,0
hypertension,0
heart_disease,0
smoking_history,0
bmi,0
HbA1c_level,0
blood_glucose_level,0
diabetes,0


In [8]:
df.describe()

,age,hypertension,heart_disease,bmi,HbA1c_level,blood_glucose_level,diabetes
count,100000.000000,100000.00000,100000.000000,100000.000000,100000.000000,100000.000000,100000.000000
mean,41.885856,0.07485,0.039420,27.320767,5.527507,138.058060,0.085000
std,22.516840,0.26315,0.194593,6.636783,1.070672,40.708136,0.278883
min,0.080000,0.00000,0.000000,10.010000,3.500000,80.000000,0.000000
25%,24.000000,0.00000,0.000000,23.630000,4.800000,100.000000,0.000000
50%,43.000000,0.00000,0.000000,27.320000,5.800000,140.000000,0.000000
75%,60.000000,0.00000,0.000000,29.580000,6.200000,159.000000,0.000000
max,80.000000,1.00000,1.000000,95.690000,9.000000,300.000000,1.000000


In [9]:
import pandas as pd

# Load dataset
df = pd.read_csv("/content/diabetes_prediction_dataset.csv")

# First 5 rows
print(df.head())

# Dataset size
print("\nDataset Shape:")
print(df.shape)

# Dataset information
print("\nDataset Information:")
df.info()

# Missing values
print("\nMissing Values:")
print(df.isnull().sum())

# Statistics
print("\nStatistical Summary:")
print(df.describe())

   gender   age  hypertension  heart_disease smoking_history    bmi  \
0  Female  80.0             0              1           never  25.19   
1  Female  54.0             0              0         No Info  27.32   
2    Male  28.0             0              0           never  27.32   
3  Female  36.0             0              0         current  23.45   
4    Male  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0  

Dataset Shape:
(100000, 9)

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   gender             

In [10]:
# Count diabetes classes
print(df["diabetes"].value_counts())

# Calculate percentage
print("\nPercentage Distribution:")
print(df["diabetes"].value_counts(normalize=True) * 100)

diabetes
0    91500
1     8500
Name: count, dtype: int64

Percentage Distribution:
diabetes
0    91.5
1     8.5
Name: proportion, dtype: float64


The dataset contains 100,000 records and 9 attributes. The target variable, diabetes, is imbalanced, with 91.5% non-diabetic records and 8.5% diabetic records. Therefore, model performance will be evaluated using metrics such as precision, recall, F1-score, and ROC-AUC rather than accuracy alone.

In [11]:
 # Check unique values in categorical columns

print("Gender:")
print(df["gender"].value_counts())

print("\nSmoking History:")
print(df["smoking_history"].value_counts())

Gender:
gender
Female    58552
Male      41430
Other        18
Name: count, dtype: int64

Smoking History:
smoking_history
No Info        35816
never          35095
former          9352
current         9286
not current     6447
ever            4004
Name: count, dtype: int64


The dataset contains two categorical features: gender and smoking history. Gender consists mainly of Female and Male categories, with a small number of Other records. Smoking history contains six categories, including No Info, which represents unknown smoking information. These categorical features will be encoded before model training.

In [12]:
# Check duplicate rows
print("Number of duplicate rows:")
print(df.duplicated().sum())

Number of duplicate rows:
3854


In [13]:
# Display some duplicate rows
df[df.duplicated()].head()

,gender,age,hypertension,heart_disease,smoking_history,bmi,HbA1c_level,blood_glucose_level,diabetes
2756,Male,80.0,0,0,No Info,27.32,6.6,159,0
3272,Female,80.0,0,0,No Info,27.32,3.5,80,0
3418,Female,19.0,0,0,No Info,27.32,6.5,100,0
3939,Female,78.0,1,0,former,27.32,3.5,130,0
3960,Male,47.0,0,0,No Info,27.32,6.0,200,0


In [14]:
# Remove duplicate rows
df = df.drop_duplicates()

# Check the new dataset shape
print("New Dataset Shape:", df.shape)

# Confirm duplicates
print("Remaining Duplicates:", df.duplicated().sum())

New Dataset Shape: (96146, 9)
Remaining Duplicates: 0


The dataset initially contained 3,854 duplicate records. These duplicate rows were removed during preprocessing, resulting in 96,146 unique records for further analysis and model development.

The dataset was checked for duplicate records. A total of 3,854 duplicate rows were identified and removed, resulting in a cleaned dataset containing 96,146 unique records.

In [15]:
# Check target distribution after removing duplicates

print(df["diabetes"].value_counts())

print("\nPercentage Distribution:")
print((df["diabetes"].value_counts(normalize=True) * 100).round(2))

diabetes
0    87664
1     8482
Name: count, dtype: int64

Percentage Distribution:
diabetes
0    91.18
1     8.82
Name: proportion, dtype: float64


After removing duplicate records, the target variable distribution was analyzed again. The cleaned dataset contains 87,664 non-diabetic records (91.18%) and 8,482 diabetic records (8.82%). The dataset therefore remains imbalanced, and evaluation metrics such as precision, recall, F1-score, and ROC-AUC will be considered during model evaluation.

In [16]:
# Separate input features and target

X = df.drop("diabetes", axis=1)
y = df["diabetes"]

print("Features Shape:", X.shape)
print("Target Shape:", y.shape)

print("\nFeatures:")
print(X.columns.tolist())

Features Shape: (96146, 8)
Target Shape: (96146,)

Features:
['gender', 'age', 'hypertension', 'heart_disease', 'smoking_history', 'bmi', 'HbA1c_level', 'blood_glucose_level']


In [17]:
# One-Hot Encode categorical columns

X_encoded = pd.get_dummies(
    X,
    columns=["gender", "smoking_history"],
    drop_first=True,
    dtype=int
)

print("Encoded Features Shape:", X_encoded.shape)

print("\nEncoded Feature Columns:")
print(X_encoded.columns.tolist())

print("\nFirst 5 Rows:")
print(X_encoded.head())

Encoded Features Shape: (96146, 13)

Encoded Feature Columns:
['age', 'hypertension', 'heart_disease', 'bmi', 'HbA1c_level', 'blood_glucose_level', 'gender_Male', 'gender_Other', 'smoking_history_current', 'smoking_history_ever', 'smoking_history_former', 'smoking_history_never', 'smoking_history_not current']

First 5 Rows:
    age  hypertension  heart_disease    bmi  HbA1c_level  blood_glucose_level  \
0  80.0             0              1  25.19          6.6                  140   
1  54.0             0              0  27.32          6.6                   80   
2  28.0             0              0  27.32          5.7                  158   
3  36.0             0              0  23.45          5.0                  155   
4  76.0             1              1  20.14          4.8                  155   

   gender_Male  gender_Other  smoking_history_current  smoking_history_ever  \
0            0             0                        0                     0   
1            0             0

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Training Features:", X_train.shape)
print("Testing Features:", X_test.shape)

print("\nTraining Target:")
print(y_train.value_counts(normalize=True) * 100)

print("\nTesting Target:")
print(y_test.value_counts(normalize=True) * 100)

Training Features: (76916, 8)
Testing Features: (19230, 8)

Training Target:
diabetes
0    91.177388
1     8.822612
Name: proportion, dtype: float64

Testing Target:
diabetes
0    91.180447
1     8.819553
Name: proportion, dtype: float64


In [19]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer

# Define categorical and numerical columns
categorical_features = [
    "gender",
    "smoking_history"
]

numerical_features = [
    "age",
    "hypertension",
    "heart_disease",
    "bmi",
    "HbA1c_level",
    "blood_glucose_level"
]

# Create preprocessing pipeline
preprocessor = ColumnTransformer(
    transformers=[
        (
            "num",
            StandardScaler(),
            numerical_features
        ),
        (
            "cat",
            OneHotEncoder(
                handle_unknown="ignore",
                drop="first"
            ),
            categorical_features
        )
    ]
)

# Fit ONLY on training data
X_train_processed = preprocessor.fit_transform(X_train)

# Transform testing data
X_test_processed = preprocessor.transform(X_test)

print("Processed Training Shape:", X_train_processed.shape)
print("Processed Testing Shape:", X_test_processed.shape)

Processed Training Shape: (76916, 13)
Processed Testing Shape: (19230, 13)


**MODEL TRAINING**

In [20]:
from sklearn.linear_model import LogisticRegression

# Create the model
logistic_model = LogisticRegression(
    max_iter=1000,
    class_weight="balanced",
    random_state=42
)

# Train the model
logistic_model.fit(X_train_processed, y_train)

print("Logistic Regression model trained successfully!")

Logistic Regression model trained successfully!


In [21]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    classification_report,
    confusion_matrix
)

# Make predictions
y_pred = logistic_model.predict(X_test_processed)

# Get prediction probabilities
y_prob = logistic_model.predict_proba(X_test_processed)[:, 1]

# Metrics
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))
print("ROC-AUC Score:", roc_auc_score(y_test, y_prob))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Accuracy: 0.8845553822152886
Precision: 0.42531356898517675
Recall: 0.8797169811320755
F1 Score: 0.5734050730207533
ROC-AUC Score: 0.960068047039606

Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.89      0.93     17534
           1       0.43      0.88      0.57      1696

    accuracy                           0.88     19230
   macro avg       0.71      0.88      0.75     19230
weighted avg       0.94      0.88      0.90     19230


Confusion Matrix:
[[15518  2016]
 [  204  1492]]


The Logistic Regression model achieved an accuracy of 88.46% and an ROC-AUC score of 96.01%. The model achieved a recall of 87.97% for the diabetic class, correctly identifying 1,492 out of 1,696 diabetic cases. However, the precision was 42.53%, indicating a relatively high number of false-positive predictions. Therefore, additional machine learning models will be trained and compared before selecting the final model.

**RANDOM_FOREST**

In [22]:
from sklearn.ensemble import RandomForestClassifier

# Create Random Forest model
rf_model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    class_weight="balanced",
    n_jobs=-1
)

# Train the model
rf_model.fit(X_train_processed, y_train)

print("Random Forest model trained successfully!")

Random Forest model trained successfully!


In [23]:
# Make predictions
y_pred_rf = rf_model.predict(X_test_processed)

# Get prediction probabilities
y_prob_rf = rf_model.predict_proba(X_test_processed)[:, 1]

# Metrics
print("Accuracy:", accuracy_score(y_test, y_pred_rf))
print("Precision:", precision_score(y_test, y_pred_rf))
print("Recall:", recall_score(y_test, y_pred_rf))
print("F1 Score:", f1_score(y_test, y_pred_rf))
print("ROC-AUC Score:", roc_auc_score(y_test, y_prob_rf))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

Accuracy: 0.9694747789911596
Precision: 0.9482619240097009
Recall: 0.691627358490566
F1 Score: 0.7998636208660075
ROC-AUC Score: 0.9628649042507174

Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98     17534
           1       0.95      0.69      0.80      1696

    accuracy                           0.97     19230
   macro avg       0.96      0.84      0.89     19230
weighted avg       0.97      0.97      0.97     19230


Confusion Matrix:
[[17470    64]
 [  523  1173]]


In [24]:
import pandas as pd
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

thresholds = [0.3, 0.4, 0.5, 0.6]

results = []

for threshold in thresholds:
    predictions = (y_prob_rf >= threshold).astype(int)

    results.append({
        "Threshold": threshold,
        "Accuracy": accuracy_score(y_test, predictions),
        "Precision": precision_score(y_test, predictions),
        "Recall": recall_score(y_test, predictions),
        "F1 Score": f1_score(y_test, predictions)
    })

threshold_results = pd.DataFrame(results)

print(threshold_results)

   Threshold  Accuracy  Precision    Recall  F1 Score
0        0.3  0.959230   0.778388  0.751769  0.764847
1        0.4  0.965627   0.873646  0.713443  0.785459
2        0.5  0.969423   0.946055  0.692807  0.799864
3        0.6  0.970723   0.980492  0.681604  0.804174


In [25]:
thresholds = [0.20, 0.25, 0.30, 0.35, 0.40, 0.45, 0.50]

results = []

for threshold in thresholds:
    predictions = (y_prob_rf >= threshold).astype(int)

    results.append({
        "Threshold": threshold,
        "Precision": precision_score(y_test, predictions),
        "Recall": recall_score(y_test, predictions),
        "F1 Score": f1_score(y_test, predictions)
    })

threshold_results = pd.DataFrame(results)

print(threshold_results)

   Threshold  Precision    Recall  F1 Score
0       0.20   0.633770  0.798939  0.706834
1       0.25   0.714208  0.773585  0.742712
2       0.30   0.778388  0.751769  0.764847
3       0.35   0.830986  0.730542  0.777534
4       0.40   0.873646  0.713443  0.785459
5       0.45   0.915765  0.698703  0.792642
6       0.50   0.946055  0.692807  0.799864


In [26]:
import xgboost
print(xgboost.__version__)

3.4.1


In [27]:
from xgboost import XGBClassifier

# Create XGBoost model
xgb_model = XGBClassifier(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=5,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    eval_metric="logloss"
)

# Train the model
xgb_model.fit(
    X_train_processed,
    y_train
)

print("XGBoost model trained successfully!")

XGBoost model trained successfully!


In [28]:
# Make predictions
y_pred_xgb = xgb_model.predict(X_test_processed)

# Get prediction probabilities
y_prob_xgb = xgb_model.predict_proba(X_test_processed)[:, 1]

# Metrics
print("Accuracy:", accuracy_score(y_test, y_pred_xgb))
print("Precision:", precision_score(y_test, y_pred_xgb))
print("Recall:", recall_score(y_test, y_pred_xgb))
print("F1 Score:", f1_score(y_test, y_pred_xgb))
print("ROC-AUC Score:", roc_auc_score(y_test, y_prob_xgb))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_xgb))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_xgb))

Accuracy: 0.9718148725949038
Precision: 0.9840604026845637
Recall: 0.691627358490566
F1 Score: 0.8123268698060941
ROC-AUC Score: 0.9775939529076662

Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98     17534
           1       0.98      0.69      0.81      1696

    accuracy                           0.97     19230
   macro avg       0.98      0.85      0.90     19230
weighted avg       0.97      0.97      0.97     19230


Confusion Matrix:
[[17515    19]
 [  523  1173]]


In [29]:
thresholds = [0.20, 0.25, 0.30, 0.35, 0.40, 0.45, 0.50]

results_xgb = []

for threshold in thresholds:
    predictions = (y_prob_xgb >= threshold).astype(int)

    results_xgb.append({
        "Threshold": threshold,
        "Accuracy": accuracy_score(y_test, predictions),
        "Precision": precision_score(y_test, predictions),
        "Recall": recall_score(y_test, predictions),
        "F1 Score": f1_score(y_test, predictions)
    })

threshold_results_xgb = pd.DataFrame(results_xgb)

print(threshold_results_xgb)

   Threshold  Accuracy  Precision    Recall  F1 Score
0       0.20  0.950806   0.687876  0.809552  0.743770
1       0.25  0.959698   0.770088  0.774175  0.772126
2       0.30  0.965367   0.847035  0.741156  0.790566
3       0.35  0.968955   0.904937  0.724057  0.804455
4       0.40  0.969891   0.933955  0.708726  0.805900
5       0.45  0.970931   0.960324  0.699292  0.809280
6       0.50  0.971815   0.984060  0.691627  0.812327


In [30]:
# Final candidate threshold
final_threshold = 0.30

# Predictions using custom threshold
y_pred_final = (y_prob_xgb >= final_threshold).astype(int)

# Confusion matrix
cm_final = confusion_matrix(y_test, y_pred_final)

print("Confusion Matrix:")
print(cm_final)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_final))

print("\nAccuracy:", accuracy_score(y_test, y_pred_final))
print("Precision:", precision_score(y_test, y_pred_final))
print("Recall:", recall_score(y_test, y_pred_final))
print("F1 Score:", f1_score(y_test, y_pred_final))
print("ROC-AUC:", roc_auc_score(y_test, y_prob_xgb))

Confusion Matrix:
[[17307   227]
 [  439  1257]]

Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.99      0.98     17534
           1       0.85      0.74      0.79      1696

    accuracy                           0.97     19230
   macro avg       0.91      0.86      0.89     19230
weighted avg       0.96      0.97      0.96     19230


Accuracy: 0.9653666146645866
Precision: 0.8470350404312669
Recall: 0.7411556603773585
F1 Score: 0.7905660377358491
ROC-AUC: 0.9775939529076662


In [31]:
import joblib

# Save the trained XGBoost model
joblib.dump(xgb_model, "medintel_xgb_model.pkl")

# Save the preprocessing pipeline
joblib.dump(preprocessor, "medintel_preprocessor.pkl")

# Save the threshold separately
joblib.dump(0.30, "medintel_threshold.pkl")

print("Model, preprocessor, and threshold saved successfully!")

Model, preprocessor, and threshold saved successfully!


In [32]:
import os

print(os.listdir())

['.config', 'medintel_xgb_model.pkl', 'diabetes_prediction_dataset.csv', 'medintel_threshold.pkl', 'medintel_preprocessor.pkl', 'sample_data']


In [33]:
import joblib
import pandas as pd

# Load saved files
model = joblib.load("medintel_xgb_model.pkl")
preprocessor_loaded = joblib.load("medintel_preprocessor.pkl")
threshold = joblib.load("medintel_threshold.pkl")

# Create sample patient data
sample_patient = pd.DataFrame({
    "gender": ["Female"],
    "age": [50],
    "hypertension": [0],
    "heart_disease": [0],
    "smoking_history": ["never"],
    "bmi": [28.5],
    "HbA1c_level": [6.2],
    "blood_glucose_level": [150]
})

# Apply the same preprocessing
sample_processed = preprocessor_loaded.transform(sample_patient)

# Get diabetes probability
probability = model.predict_proba(sample_processed)[0][1]

# Apply our threshold
prediction = 1 if probability >= threshold else 0

print("Diabetes Risk Probability:", round(probability * 100, 2), "%")

if prediction == 1:
    print("Result: Higher Risk")
else:
    print("Result: Lower Risk")

Diabetes Risk Probability: 5.18 %
Result: Lower Risk


In [34]:
sample_patient_high_risk = pd.DataFrame({
    "gender": ["Male"],
    "age": [65],
    "hypertension": [1],
    "heart_disease": [1],
    "smoking_history": ["former"],
    "bmi": [32.0],
    "HbA1c_level": [7.5],
    "blood_glucose_level": [220]
})

sample_processed = preprocessor_loaded.transform(sample_patient_high_risk)

probability = model.predict_proba(sample_processed)[0][1]

prediction = 1 if probability >= threshold else 0

print("Diabetes Risk Probability:", round(probability * 100, 2), "%")

if prediction == 1:
    print("Result: Higher Risk")
else:
    print("Result: Lower Risk")

Diabetes Risk Probability: 99.99 %
Result: Higher Risk
